### PR TITLE
Type ignore comments erroneously marked as unused by dmypy

### PR DESCRIPTION
**Pull request by meshy**
There is currently a misbehaviour where "type: ignore" comments are erroneously marked as unused in re-runs of dmypy. There are also cases where errors disappear on the re-run.

As far as I can tell, this only happens in modules which contain an import that we don't know how to type (such as a module which does not exist), and a submodule which is unused.

There was a lot of commenting and investigation on this PR, but I hope that the committed tests and fixes illustrate and address the issue.

Related to https://github.com/python/mypy/issues/9655

### CODE REVIEW

**Comment by chadrik** [comment]
I ran into this issue testing a pre-release of 1.3 and it has created enough confusion on the team that I'm not sure we can roll it out until this is fixed.  I would love to see this merged before release.

**Comment by meshy** [comment]
I have discovered that this test passes when I revert https://github.com/python/mypy/pull/14835, so it's probably fair to consider this a regression. I haven't yet discovered the mechanism behind this new failure.

Conversely, https://github.com/python/mypy/pull/15049 is not fixed by reverting my last PR.

**Comment by meshy** [comment]
Once https://github.com/python/mypy/pull/15179 reverts my previous PR, I will rebase this to include those original commits, and hope to come up with a solution that will make both the old and new tests pass.

**Comment by meshy** [comment]
I've rebased again to get my tests running smoothly against the latest version of the code.

CI still fails on the regression test added in the last commit.

**TLDR:** I've finally tracked down the cause of that failure, but my current fix breaks loads of other stuff, so needs more work.

What follows is my best understanding of the issue as it stands. Please excuse me if I waffle a bit.

So as a recap, the setup for the remaining failing test looks like this:

```
[file unused/__init__.py]
[file unused/empty.py]
[file foo.py]
from unused.empty import *
import a_module_which_does_not_exist
def is_foo() -> str:
    return True  # type: ignore
```

The first run of `dmypy` works correctly and has this (fairly boring) output:

```
foo.py:2: error: Cannot find implementation or library stub for module named "a_module_which_does_not_exist"  [import-not-found]
foo.py:2: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

But the second run adds this extra error on the end which is clearly wrong:

```
foo.py:4: error: Unused "type: ignore" comment  [unused-ignore]
```

I believe this is because in the first run, when the ignored error is recorded we return without adding the error into `error_info_map`. https://github.com/python/mypy/blob/10dfafe089a75dc117586ebab35723da66309398/mypy/errors.py#L501-L506

Then in the second run when we fetch the list of things to re-check, `foo.is_foo` is missing from `previous_targets_with_errors`. https://github.com/python/mypy/blob/10dfafe089a75dc117586ebab35723da66309398/mypy/server/update.py#L189

... which is populated from the `error_info_map` (notice the comment here, it's relevant later). https://github.com/python/mypy/blob/10dfafe089a75dc117586ebab35723da66309398/mypy/errors.py#L929-L935

I have found that changing `add_error_info` so that it adds this ignored error into `error_info_map` (with `info.hidden=True` does indeed solve the spurious `Unused "type: ignore" comment` error.

```diff
diff --git mypy/errors.py mypy/errors.py
index 4cdeb54f8..0fa365682 100644
--- mypy/errors.py
+++ mypy/errors.py
@@ -503,6 +503,9 @@ class Errors:
                         self.used_ignored_lines[file][scope_line].append(
                             (info.code or codes.MISC).code
                         )
+                        if file not in self.ignored_files:
+                            info.hidden = True
+                            self._add_error_info(file, info)
                         return
             if file in self.ignored_files:
                 return
```

Unfortunately, it also causes many other tests to fail. Some have errors like this:

`mypy/test/testcheck.py::TypeCheckSuite::check-inline-config.test::testInlineIncremental1`:
```
E   AssertionError: Some modules reported error despite no errors
E   assert (set() or not {'/home/charlie/code/mypy/test-data/unit/lib-stub/typing.pyi', 'main', 'tmp/a.py', 'tmp/builtins.pyi'})
```

I might be able to fix those by changing the helper functions on the `Errors` to filter out `hidden` errors, or add a new `ignored` flag on `ErrorInfo` for the same purpose. I haven't tried this yet. I'm hoping to find a more elegant solution.

Others fail like this:

```
Expected:
  ==
Actual:
  ==
  builtins.pyi:18: error: Name "Iterator" is not defined (diff)
  builtins.pyi:18: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Iterator") (diff)
```

... and some fail like this:
Expected:
  ==
  main:4: error: Incompatible return value type (got "str", expected "int")
Actual:
  ==
  main:4: error: Incompatible return value type (got "str", expected "int")
  builtins.pyi:6: error: Name "typing" is not defined (diff)
```

Perhaps the fix for these errors would be to make sure that `ErrorInfo`'s `.target` is always set to a dotted-Python-path, so that they're caught by `ignored_files`. I haven't tried that out yet (or even proved the theory) but I think that's the next thing to try. I think this is supported by the previously highlighted comment: "TODO: Make sure that either target is always defined or that not being defined is okay for fine-grained incremental checking". I think it's not OK, and needs to be always defined.

**Comment by meshy** [comment]
Recap: the patch in the comment above fixed the tested error, but caused (many) other tests to fail.

I now think that my theory from the previous comment (final paragraph) was wrong.

With @samueljsb's assistance, I've come to the conclusion that the patch is correct but might be revealing a previously unnoticed issue with the tests.

Example: `mypy/test/testfinegrained.py::FineGrainedSuite::fine-grained.test::testAttrsUpdate2` fails with the patch applied with this error:

```
_____ testAttrsUpdate2 _____
[gw0] linux -- Python 3.10.12 /home/charlie/venvs/mypy/bin/python
data: /home/charlie/code/mypy/test-data/unit/fine-grained.test:940:
Failed: Invalid output (/home/charlie/code/mypy/test-data/unit/fine-grained.test, line 940)
----- Captured stderr call -----
Expected:
  ==
  main:2: error: Missing positional argument "b" in call to "B"
Actual:
  ==
  main:2: error: Missing positional argument "b" in call to "B"
  builtins.pyi:18: error: Name "Iterator" is not defined (diff)
  builtins.pyi:18: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Iterator") (diff)
```

But applying this further patch fixes the test:

```diff
diff --git test-data/unit/fine-grained.test test-data/unit/fine-grained.test
index 68f72a2aa..1977f709e 100644
--- test-data/unit/fine-grained.test
+++ test-data/unit/fine-grained.test
@@ -959,6 +959,7 @@ class A:
     a = attr.ib()   # type: int
     other = attr.ib()   # type: int
 [builtins fixtures/list.pyi]
+[typing fixtures/typing-namedtuple.pyi]
 [out]
 ==
 main:2: error: Missing positional argument "b" in call to "B"
```

This leaves rather a lot of other failing tests yet to fix before we can merged. I haven't yet proved that all the other failing tests share this solution.

After applying the main patch and running against my codebase I still see that dmypy's second run's results differ from the first run's. I assume that is because of the issue outlined in https://github.com/python/mypy/pull/15049, which I am yet to fix.

**Comment by meshy** [comment]
I have found a nicer solution to the failing tests.

I have found that changing `is_errors_for_file` to ignore hidden messages keeps the tests passing when the patch which adds ignored errors is applied.

**Comment by meshy** [comment]
Many of the failing tests relate to `Iterator` not being defined. I've found that use of this pattern in the files in `test-data/unit/fixtures/` seems to be common to the cases I've looked at:

```python
class Iterable(Protocol[T_co]):
    def __iter__(self) -> 'Iterator[T_co]': pass

class Iterator(Iterable[T_co], Protocol):
    def __next__(self) -> T_co: pass
```

I notice that `Iterable` is defined in terms of `Iterator`, and `Iterator` is defined in terms of `Iterable`. I would have expected that the quotes in the definition of `__iter__` should be enough to make this a legal construction, but removing the `__iter__` definition makes many of the tests pass. Unfortunately, it also makes many others fail.

I'm not yet sure if the solution here is to change how Iterable is defined, or to work out why the quotes aren't doing the trick. If anyone has any insight on this issue, I'd appreciate the assistance!

**Comment by JelleZijlstra** [comment]
The quotes shouldn't matter as this is a stub file, which isn't evaluated. The mutually recursive definition of Iterable and Iterator is indeed annoying to deal with, but mypy should be able to handle it.

**Comment by meshy** [comment]
I'm pretty sure that these stub files are handled in the tests, especially where we override `test-data/unit/lib-stub/typing.pyi` with declarations like this:
https://github.com/python/mypy/blob/2aa2443107534715a650dbe78474e7d91cc9df20/test-data/unit/check-incremental.test#L4154-L4155

https://github.com/python/mypy/blob/2aa2443107534715a650dbe78474e7d91cc9df20/test-data/unit/fixtures/typing-medium.pyi#L3-L4

I've found that using simpler stub files that lack this circular definition works in the case of many tests, so it feels like there must be some kind of evaluation going on.

**Comment by seddonym** [comment]
I'd like to try to help with this issue, and have some time next week.

From what I understand, the PR fixes the underlying issue, but causes some apparently unrelated tests to fail in a surprising way. (The odd behaviour is limited to tests and doesn't happen if we run mypy for real.)

I'm assuming this is a problem with the test machinery, and in particular with test fixtures that attempt to import `Iterator`.

My plan next week is to understand better what's going on by stepping through the test run in a debugger, but if anyone has any inkling as to what might be going on I'd be very grateful for any pointers. @chadrik or @JelleZijlstra I don't suppose you have any idea?

**Comment by seddonym** [comment]
I've created a simple test case to show what is breaking. This test passes (but it shouldn't).

```
[case testDemo]
[file a.py]
from typing import List
a: List[int] = [1]
[file a.py.2]
from typing import List
a: List[int] = [1]
# Comment to trigger reprocessing.
[builtins fixtures/list.pyi]
[out]
==
builtins.pyi:18: error: Name "Iterator" is not defined
builtins.pyi:18: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Iterator")
```

This test is simulating the addition of a comment to a file that is using something from the `typing` module (in this case `List`). That means we need to use the `list.pyi` fixture. The first time the module is checked (above the `==`) there are no errors, but the second time (below the `==`) we get the `"Iterator" is not defined` error.

I'll continue to try to understand why this is...

**Comment by seddonym** [comment]
Good news - I think I've got somewhere. [This WIP commit](https://github.com/python/mypy/pull/16675/commits/4c37dac8f2725a2377e3454c9bcb1b6fb729203a) gets the number of failing tests down to only 3. 🥳 

# Background

I noticed while debugging that there were some additional errors in the error list following the first check run. I found these by putting a breakpoint in the final statement of `mypy.dmypy_server.Server.initialize_fine_grained` and then drilling down to 
`self.fine_grained_manager.manager.errors.error_info_map`.

Most of the errors were along these lines:

```
'Method "__getitem__" is not using @override but is overriding a method in class "typing.Sequence"'
'Method "__iter__" is not using @override but is overriding a method in class "typing.Iterable"'
```

# How I fixed it

I'm not sure if it's the correct thing to do, but I found adding `from typing_extensions import override` and then decorating the methods in question made the tests pass.

@chadrik @JelleZijlstra Do you have any concerns with this as a fix? If so, please advise what would be a better thing to do.

There were a couple of other test cases where instead the fix seemed to be just to add the appropriate `builtins` declaration in the test case.

# Remaining failures

The last three are a bit different - I'll continue to look at these tomorrow.

**Comment by tmke8** [comment]
Sounds like the error code `explicit-override` from #15512 is being activated erroneously in the test stubs. Adding the `@override` decorators feels more like a work-around for that.

**Comment by seddonym** [comment]
> Sounds like the error code explicit-override from https://github.com/python/mypy/pull/15512 is being activated erroneously in the test stubs. Adding the @override decorators feels more like a work-around for that.

Thanks for the steer, yes that makes sense.

I've taken a different route and followed the logic around error reporting. I noticed a difference between the way other typesheds are handled and specifically the `/tmp/builtins.pyi` used by the test suite. That builtins is, possibly, special and probably needs handling as a special case. By way of a demo, this [rather hacky commit](https://github.com/python/mypy/pull/16675/commits/84271b063558b9c5481aa236190c42038a9b039c) adds a couple of special casing for builtins which gets the tests passing (well, the same tests as adding those `overrides` does). I doubt this is mergeable in its current form but perhaps it might give someone more familiar with the code base an idea of what the right fix might be?

**Comment by seddonym** [comment]
I've found that I can get it down to one failing test with a couple of less-hacky feeling changes, see the commits on the end of [this draft PR](https://github.com/python/mypy/pull/16681/commits).

The main difference is that we filter by hidden targets when returning files with errors in `mypy.errors.Errors.targets`. However that does cause one of the new test cases to fail again.

```
mypy/test/testdaemon.py::DaemonSuite::daemon.test::testReturnTypeIgnoreAfterUnknownImport
```

My guess is that `mypy.errors.Errors.targets` should sometimes return hidden files and sometimes not, it depends on the context. Maybe we should have two methods instead of one. I'm finding this harder to debug though as the failing test is in daemon.test which doesn't play as nicely with my interactive debugger.

Would welcome any thoughts from those more familiar with the internals of the daemon.

**Comment by meshy** [comment]
I've had a quick call with @seddonym (thank you for your engagement with this), and we've worked out that this patch was recording errors in too many cases. While we wanted to keep track of ignored lines, the previous changes were also tracking errors with ignored error-codes, which explains why the tests were getting so upset. This latest change greatly reduces the number of failed tests, and I've included @seddonym as a co-author accordingly.

Thanks also to @tmke8 for pointing us at the fact that this error was being registered in error, as that really helped us to track this down.

When run locally I still get a couple of wheel-related errors. They're up next!

**Comment by meshy** [comment]
Now that tests are passing, I've reworked the commit history so that dead-ends are removed, and they follow the order:

- Tests
- Refactoring
- Changes

I have a feeling that while this does pass the new tests that there may yet be some undiscovered edge-cases. I'm going to dig in and see what I can find before opening this for review. It should be noted that these changes do not yet fix the tests in https://github.com/python/mypy/pull/15049.

**Comment by meshy** [comment]
I've found that this branch now has a regression similar to the last one we found.

<details>
<summary>
Expand this for details on how to reproduce the issue.
</summary>

I can reproduce this one with these files:

`unused/init.py`: (empty)
`unused/empty.py`: (empty)
`foo.py`:
```python
import attr
from unused.empty import *
import a_module_which_does_not_exist

@attr.frozen
class A:
    def __init__(self) -> None:
        self.__attrs_init__()  # type: ignore[attr-defined]
```

Run these commands:
```
dmypy start -- --warn-unused-ignores --no-error-summary
dmypy check -- foo.py
dmypy check -- foo.py
```
Run 1 results:
```
foo.py:3: error: Cannot find implementation or library stub for module named "a_module_which_does_not_exist"  [import-not-found]
foo.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

Run 2 results:
```
foo.py:3: error: Cannot find implementation or library stub for module named "a_module_which_does_not_exist"  [import-not-found]
foo.py:3: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
foo.py:9: error: Unused "type: ignore" comment  [unused-ignore]
```
</details>

I haven't yet worked out how to make a regression test for this case in this PR yet, not investigated the cause.

This appears to be related to what I suspect may be [an issue with the auto-detection of `__init__` methods in the attrs plugin](https://github.com/python/mypy/pull/16704). I have created a test to show that issue in the linked PR.

**Comment by meshy** [comment]
This latest patch fixes all the errors I am aware of, including those in https://github.com/python/mypy/pull/15049.

My local code-base can now re-run dmypy without any changes between runs! :tada: 

Before asking for review, I want to work out if we need all of the changes I've made here, or if the latest fix ("[Don't remove modules with descendants](https://github.com/python/mypy/pull/15043/commits/a4a12dfc6f67d343e648ae1661dbccf89a640128)") suffices.

I'll merge in the tests from https://github.com/python/mypy/pull/15049 soon.

**Comment by meshy** [comment]
I've now boiled this down the the minimal changes to fix the issue (though there are a lot of tests), and I think that this is finally ready for review!

File: mypy/dmypy_server.py [hunk 1]
```diff
@@ -715,6 +716,29 @@ def refresh_file(module: str, path: str) -> list[str]:
 
         return messages
 
+    def _seen_and_ancestors(self, seen: set[str]) -> set[str]:
+        """Return the set of seen modules along with any ancestors not already in the set.
```
Comment (seddonym) [inline mypy/dmypy_server.py:720]: Great docstring, thanks.

File: mypy/dmypy_server.py [hunk 2]
```diff
@@ -715,6 +716,29 @@ def refresh_file(module: str, path: str) -> list[str]:
...
+            {"foo", "foo.bar", "a.b.c"}
+
+        ... we would expect this set to be returned:
+
+            {"foo", "foo.bar", "a.b.c", "a.b", "a"}
+
+        This is used to stop us from deleting ancestor modules from the graph
+        when their descendants have been seen.
+        """
+        seen_paths = seen.copy()
```
Comment (seddonym) [inline mypy/dmypy_server.py:733]: No big deal, but we could do this without any conditionals like this, if you prefer:
```
    seen_and_ancestors: set[str] = set()
    for module in seen:
        module_parts = module.split(".")
        for i in range(len(module_parts)):
            seen_and_ancestors.add(".".join(module_parts[:i + 1]))
    return seen_and_ancestors
```

File: mypy/dmypy_server.py [hunk 3]
```diff
@@ -715,6 +716,29 @@ def refresh_file(module: str, path: str) -> list[str]:
...
+        for module_path in seen:
+            while module_path := module_path.rpartition(".")[0]:
+                if module_path in seen_paths:
```
Comment (seddonym) [inline mypy/dmypy_server.py:736]: I think we might be able to drop this conditional and just always add the module_path.

**Review by seddonym — COMMENTED** [review]
Heroic effort @meshy!

Just left a couple of tiny comments, feel free to disregard. I'll leave approval to a maintainer!

File: mypy/dmypy_server.py (hunk 2 above)
Comment (meshy) [reply mypy/dmypy_server.py:733]: I assumed that in some cases with many sub-modules that my way would be faster so my preference would be to keep it as-is, but I'm not strongly opposed to changing this.

File: mypy/errors.py [hunk 4]
```diff
@@ -501,19 +501,29 @@ def add_error_info(self, info: ErrorInfo) -> None:
...
                             info.code.code in ignores[scope_line]
                             or info.code.sub_code_of is not None
                             and info.code.sub_code_of.code in ignores[scope_line]
                         )
+                        record_ignored_line = is_ignored_error
                     else:
                         is_ignored_error = False
+                        record_ignored_line = False
+
+                    if record_ignored_line and file not in self.ignored_files:
```
[the remark this answers is not in the document]
Comment (meshy) [reply mypy/errors.py]: This code isn't in the PR any more

File: mypy/dmypy_server.py (hunk 3 above)
Comment (meshy) [reply mypy/dmypy_server.py:736]: It feels like it'd be quicker to skip this if we have a module with many submodules.

**Comment by CoolCat467** [comment]
Testing out installing this branch of your repo locally, I found the repeat dmypy runs on a module I installed in edit mode (`pip install -e <module>`) were still ignoring errors, and then attempting to preform a dmypy run on another file that used that module resulted in a crash.

<details>

> dmypy --status-file="<my_home_folder>/.idlerc/mypy/dmypy.json" run --timeout=1800 --log-file="<my_home_folder>/.idlerc/mypy/log.txt" --export-types "<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py" -- --warn-redundant-casts --disallow-untyped-calls --cache-dir="<my_home_folder>/.idlerc/mypy" --cache-fine-grained --warn-unused-ignores --show-traceback --no-error-summary --show-column-numbers --no-color-output --show-absolute-path --hide-error-context --strict --warn-unreachable --soft-error-limit=-1 --disallow-untyped-defs --show-error-end --no-implicit-reexport --show-error-codes


<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py:539:9:539:27: error: Property "enabled" defined in "Connection" is read-only  [misc]
<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py:658:13:658:30: error: Property "enabled" defined in "Connection" is read-only  [misc]
<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py:745:13:745:30: error: Property "enabled" defined in "Connection" is read-only  [misc]
<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py:933:38:933:38: error: Cannot use a covariant type variable as a parameter  [misc]
<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py:1320:26:1320:35: error: Only concrete class can be given where "type[BasePlayer[Any]]" is expected  [type-abstract]
<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py:1323:37:1323:46: error: Only concrete class can be given where "type[BasePlayer[Any]]" is expected  [type-abstract]


> dmypy --status-file="<my_home_folder>/.idlerc/mypy/dmypy.json" run --timeout=1800 --log-file="<my_home_folder>/.idlerc/mypy/log.txt" --export-types "<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/src/neat/neat.py" -- --warn-redundant-casts --disallow-untyped-calls --cache-dir="<my_home_folder>/.idlerc/mypy" --cache-fine-grained --warn-unused-ignores --show-traceback --no-error-summary --show-column-numbers --no-color-output --show-absolute-path --hide-error-context --strict --warn-unreachable --soft-error-limit=-1 --disallow-untyped-defs --show-error-end --no-implicit-reexport --show-error-codes


No content returned


> dmypy --status-file="<my_home_folder>/.idlerc/mypy/dmypy.json" run --timeout=1800 --log-file="<my_home_folder>/.idlerc/mypy/log.txt" --export-types "<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/examples/minimax/checkers_ai.py" -- --disallow-untyped-defs --hide-error-context --no-implicit-reexport --no-color-output --show-error-end --show-absolute-path --cache-fine-grained --no-error-summary --soft-error-limit=-1 --warn-unused-ignores --disallow-untyped-calls --warn-unreachable --show-column-numbers --show-traceback --warn-redundant-casts --cache-dir="<my_home_folder>/.idlerc/mypy" --strict --show-error-codes


Daemon crashed!
Traceback (most recent call last):
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/dmypy_server.py", line 236, in serve
    resp = self.run_command(command, data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/dmypy_server.py", line 285, in run_command
    ret = method(self, **data)
          ^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/dmypy_server.py", line 353, in cmd_run
    return self.check(sources, export_types, is_tty, terminal_width)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/dmypy_server.py", line 429, in check
    messages = self.fine_grained_increment_follow_imports(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/dmypy_server.py", line 631, in fine_grained_increment_follow_imports
    messages = fine_grained_manager.update(changed, [], followed=True)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/server/update.py", line 267, in update
    result = self.update_one(
             ^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/server/update.py", line 369, in update_one
    result = self.update_module(next_id, next_path, next_id in removed_set, followed)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/server/update.py", line 431, in update_module
    result = update_module_isolated(
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/server/update.py", line 670, in update_module_isolated
    state.generate_unused_ignore_notes()
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/build.py", line 2583, in generate_unused_ignore_notes
    self.manager.errors.generate_unused_ignore_errors(self.xpath)
  File "<my_home_folder>/.local/lib/python3.12/site-packages/mypy/errors.py", line 690, in generate_unused_ignore_errors
    if line in self.skipped_lines[file]:
               ~~~~~~~~~~~~~~~~~~^^^^^^
KeyError: '<my_home_folder>/Desktop/Python/Ports/NEAT-Template-Python/examples/minimax/checkers_ai.py'
</details>

**Comment by meshy** [comment]
@CoolCat467 That's great info, thank you very much.

Have you tested that crash on Mypy's `master` branch? Is it definitely a problem introduced by these changes?

I'll see if I can reproduce those disappearing errors.

If you're able to produce a minimal example of any of these failures I'd really appreciate it!

**Comment by meshy** [comment]
@CoolCat467 I've reproduced a crash on the `master` branch using your example. The crash I got wasn't exactly the same, but it's possible that's because I didn't have exactly the same setup as you.

I don't think that the crash I got is related to these changes, though as I said, I didn't get the same one as you.

I've boiled it down to a minimal example, and added details of the crash in https://github.com/python/mypy/issues/14645#issuecomment-1881493341

**Comment by meshy** [comment]
@CoolCat467 I'm not able to reproduce your issue with the disappearing messages using the script in that repo.

Can I please ask for detailed instructions on how to reproduce it?

**Comment by CoolCat467** [comment]
> @CoolCat467 I'm not able to reproduce your issue with the disappearing messages using the script in that repo.
> 
> Can I please ask for detailed instructions on how to reproduce it?

I have now updated `https://github.com/CoolCat467/NEAT-Template-Python/dmypy_test.sh` to use absolute paths and I get the key error reliably. I would be happy to help if you need more details!

**Comment by meshy** [comment]
@CoolCat467 Thank you for the clarification. I've got that working now, and can reproduce the dropped errors (on this branch and master). I'll try to turn this into a minimal test case.

Curiously, I'm now not seeing the crash on the third command, but I'll come back to that later.

**Comment by meshy** [comment]
The best minimal example I have so far:

`pyproject.toml`:
```toml
[build-system]
requires = ["setuptools >= 64"]
build-backend = "setuptools.build_meta"

[project]
name = "dmypy_example"
version = "0.0.1"
dependencies = []
```

`src/foo.py` (see notes at end):
```python
a: str = 1
```

Commands:

```sh
# Create a new Python 3.12 virtualenv, then:
pip install --editable .
dmypy run --export-types $(pwd)/src/foo.py
dmypy run --export-types $(pwd)/src/foo.py
```

Results:
```console
$ pip install --editable .
...

$ dmypy run --export-types $(pwd)/src/foo.py
Daemon started
src/foo.py:1: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
Found 1 error in 1 file (checked 1 source file)

$ dmypy run --export-types $(pwd)/src/foo.py
Success: no issues found in 1 source file
```

Notes:

- If I use a name other than `src` this doesn't seem to work. This is so odd that I feel like I must be doing something wrong.
- This doesn't happen if we don't `pip install`.
- It doesn't matter what the error in `foo.py` is. All errors seem to disappear in the second run.
- This happens on the `master` branch of Mypy too

My conclusion:

It's disappointing that this branch doesn't fix this error, but I don't think that it's _caused_ by this branch, so perhaps we should consider this a new bug for handling separately?

**Comment by CoolCat467** [comment]
> It's disappointing that this branch doesn't fix this error, but I don't think that it's caused by this branch, so perhaps we should consider this a new bug for handling separately?

Yea, I think it would make sense to make a new issue for this in particular.

> If I use a name other than `src` this doesn't seem to work. This is so odd that I feel like I must be doing something wrong.

I think this might be due to the project structure pip expects, but not completely sure

Also side note, thank you for showing the use of `pwd`, I wasn't aware that was a thing, that's super helpful instead of using python to figure out the path!

**Comment by meshy** [comment]
> > If I use a name other than `src` this doesn't seem to work. This is so odd that I feel like I must be doing something wrong.
> 
> I think this might be due to the project structure pip expects, but not completely sure

As far as I know, `src` is nothing to do with `pip`, it's a convention for structuring code that has been gaining momentum in the last few years. (Not a good one, in my opinion, but that's off topic.)

> thank you for showing the use of `pwd`

I'm glad you found it useful! :slightly_smiling_face: 

> > It's disappointing that this branch doesn't fix this error, but I don't think that it's caused by this branch, so perhaps we should consider this a new bug for handling separately?
> 
> Yea, I think it would make sense to make a new issue for this in particular.

Thanks! I've opened a new issue so we can track it there.  https://github.com/python/mypy/issues/16768

**Comment by meshy** [comment]
Hi all! It's been a while since I've had a look at this (I got a bit overwhelmed with it, TBH).

I notice that this was considered for release in 1.9, but wasn't merged because ["while the PR itself is apparently ok, it doesn't actually fix the issue completely that its trying to address"](https://github.com/python/mypy/issues/16843#issuecomment-1930629142). 

Am I right in saying that [the issue discovered in this comment](https://github.com/python/mypy/issues/9655#issuecomment-1927400546) is the blocker? Before I go about investigating that, are there any other blockers that we're aware of?

**Comment by meshy** [comment]
@jhance I wonder if you're able to help me answer [the question above](https://github.com/python/mypy/pull/15043#issuecomment-2324387438), so that I might be able to continue on this?

**Review by ilevkivskyi — APPROVED** [review]

**Comment by meshy** [comment]
OMG you merged it! I love you!
